feat: Update table view layout according to design

andreaforni opened the pull request:
This PR closes #82 and #115

The PR introduces the material UI components in the "Overall schema" panel and:
- adds column sorting to all the tables.
- removes the `SASS` package from the project because all the `scss` files have been removed.
- saves the open accordions and subtables in the local storage (using the `PluginState` context).

![Screenshot 2022-12-15 at 19 07 49](https://user-images.githubusercontent.com/1233582/207935416-ead99530-523a-4125-87d9-a7d78bc9631a.png)


## How to test it

To test this PR:
- install the dependencies with `npm ci` 
- execute the command `npm run example` 
- open your browser to http://localhost:3001/graphiql
- then click on the plugin tab

Please note that this solution does not require the `mercurius-federation-info` project to run because the plugin example builds the source code and runs a mercurius instance for you.

andreaforni commented:
> Dark theme doesn't look right: ![chrome_wM4ZiljVHU](https://user-images.githubusercontent.com/14069265/207830472-e92d17bb-ad9c-4043-a32f-528ff63bb97f.png)

I know the tables look too black in the dark mode, but this is the default MUI dark theme. If you go to the [MUI table doc](https://mui.com/material-ui/react-table/) and switch to the dark mode; you get the same colour:

![Screenshot 2022-12-15 at 11 32 35](https://user-images.githubusercontent.com/1233582/207837370-4263a8f9-471e-43fb-80c1-c2dc7ad1211f.png)

simoneb commented:
Just some general comments as I haven't looked much at this plugin before.

1. we could use some padding around items in the tree

![image](https://user-images.githubusercontent.com/20181/207839845-5712e155-6b95-448f-8039-ce5cdb0ef721.png)

2. "Overall Schema" should probably be "Unified Schema" or something like that. See https://principledgraphql.com/integrity#1-one-graph

3. Service names are with initial upper case in the tree view but lowercase in the tables

4. I don't find particularly intuitive that the return types of e.g. queries or the types of fields in the tree view are nested children of the node. I understand it's not easy to capture them otherwise, but I'd give it some thought

5. the UI looks very "big", I'd consider making it a little more compact (iow, its seems it's using lots of screen space for no good reasons)

matijagaspar commented on src/federation-info-plugin/views/SchemaView/SchemaView.jsx:
```diff
@@ -89,38 +61,61 @@ const SchemaView = ({ schemaViewData, rootTypes }) => {
     return { queries, mutations, subscriptions, types }
   }, [schemaViewData, rootTypes])
 
+  const sortQueries = (property, order) =>
```
I have mixed feelings about this way of handling the sorting, it feels a bit convoluted, and it also has a very bad performance impact (although at this level it shouldn't be noticable yet)

Could we maybe make a util function that we use directly inside `SchemaFieldsTable` and possibly memoize it there?

So instead of passing `onSortChange={sortQueries}` we use util function `SchemaFieldsTable`.

andreaforni replied:
Commit pushed based on the solution we discussed on the call.

matijagaspar commented on src/federation-info-plugin/components/SchemaFieldsTable/SchemaFieldsTable.jsx:
```diff
@@ -0,0 +1,121 @@
+  return (
+    <SchemaOperationTable
+      {...rest}
+      headerRender={({ showReference }) => (
```
I think we could get away not passing the `showReference` in this function, but we have it in the props passed to `SchemaFieldsTable` (currently inside the `...rest`).  Also no need for the headerRender to even be a function

Basically something like this:

```jsx
const SchemaFieldsTable = ({ onSortChange, showReference,  ...rest }) => {
  const [order, setOrder] = useState('desc')
  const [orderBy, setOrderBy] = useState(null)

  const createSortHandler = property => () => {
    const isAsc = orderBy === property && order === 'asc'
    setOrder(isAsc ? 'desc' : 'asc')
    setOrderBy(property)

    onSortChange && onSortChange(property, order)
  }

  return (
    <SchemaOperationTable
      {...rest}
      headerRender={
        <FieldHeader
          showReference={showReference}
          order={order}
          orderBy={orderBy}
          createSortHandler={createSortHandler}
        />
      }
      rowRender={({ field }) => (
        <FieldRow
          key={field.name}
          field={field}
          showReference={showReference}
        />
      )}
    />
  )
}
```

We could technically "pre-generate" and memoize rowRenderer, but I don't think is worht the hassle for now

andreaforni replied:
Commit pushed based on the solution we discussed on the call.

matijagaspar commented on src/federation-info-plugin/components/SchemaFieldsTable/SchemaFieldsTable.jsx:
```diff
@@ -0,0 +1,121 @@
+} from '../../utils/schemaFieldToTableCellValue'
+
+const FieldHeader = ({ showReference, order, orderBy, createSortHandler }) => {
+  return (
```
memoize all the: `createSortHandler ('field')` here I think would be the most efficent.

andreaforni replied:
Commit pushed based on the solution we discussed on the call.

andreaforni commented:
> > Dark theme doesn't look right: ![chrome_wM4ZiljVHU](https://user-images.githubusercontent.com/14069265/207830472-e92d17bb-ad9c-4043-a32f-528ff63bb97f.png)

@matijagaspar I've created a custom dark mode theme and wrapped all the theme management into the new `PluginThemeProvider `.

The light theme is the same, the dark mode is:
![Screenshot 2022-12-15 at 18 08 26](https://user-images.githubusercontent.com/1233582/207924800-c3a1eaf3-5aa0-4750-9ab1-69d9741ad246.png)

andreaforni commented:
> Just some general comments as I haven't looked much at this plugin before.
> 
> 1. we could use some padding around items in the tree
> 
> ![image](https://user-images.githubusercontent.com/20181/207839845-5712e155-6b95-448f-8039-ce5cdb0ef721.png)
> 
> 2. "Overall Schema" should probably be "Unified Schema" or something like that. See https://principledgraphql.com/integrity#1-one-graph
> 3. Service names are with initial upper case in the tree view but lowercase in the tables
> 4. I don't find particularly intuitive that the return types of e.g. queries or the types of fields in the tree view are nested children of the node. I understand it's not easy to capture them otherwise, but I'd give it some thought
> 5. the UI looks very "big", I'd consider making it a little more compact (iow, its seems it's using lots of screen space for no good reasons)

Hi @simoneb, thanks for the feedback. I've discussed them with @matijagaspar:
1. We can improve that, but I created a new issue (#115) because this PR is focused on the schema panel.
2. Changed to "Unified schema"
3. We decided to keep the original case in the service name so you will find them in lowercase. We also changed the name of the services adding the "service" suffix to distinguish them from the `User` and `Post` types.
4. Moved to issue #115 
5. Yeah, MUI generally adds a lot of space, but I've changed the table to the `condensed` mode, so they take less space.

I'll start working on #115. I've also created #116 to keep track of a "nice to have" feature that can improve the readability of the service panel when there are long names.

simoneb commented:
@andreaforni thanks!

andreaforni commented:
> @andreaforni thanks!

No problem! I've just pushed a commit to close #115. It was a quick fix.

What I've done:
- I moved the type one level up, in the same node of the field name
- I checked the node padding, and there was already an `8px` vertical padding. Adding more than `8px` adds too much white space, but to improve it a little bit the experience I moved the padding inside the content node, so when the user `hover` over a node, the grey area has some padding.

![Screenshot 2022-12-15 at 18 55 58](https://user-images.githubusercontent.com/1233582/207934079-5cf5d56c-7332-476c-b61c-dcb209e742c1.png)
![Screenshot 2022-12-15 at 18 56 14](https://user-images.githubusercontent.com/1233582/207934083-c19564db-b91d-4391-9e4c-c9aec134216a.png)

matijagaspar reviewed (approved):
LGTM
